test/v2/services/RetrieveHistoricNonFhlUkPropertyPeriodSummaryServiceSpec.scala

```diff
@@ -0,0 +1,92 @@
+import v2.models.response.retrieveHistoricNonFhlUkPiePeriodSummary.{ PeriodExpenses, PeriodIncome, RetrieveHistoricNonFhlUkPiePeriodSummaryResponse }
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+class RetrieveHistoricNonFhlUkPropertyPeriodSummaryServiceSpec extends UnitSpec {
+
+  val nino: String                   = "AA123456A"
+  val from: String                   = "2017-04-06"
+  val to: String                     = "2017-07-04"
+  val periodId: String               = s"${from}_$to"
+  implicit val correlationId: String = "X-123"
+
+  val periodIncome: PeriodIncome     = PeriodIncome(None, None, None, None, None, None)
+  val periodExpenses: PeriodExpenses = PeriodExpenses(None, None, None, None, None, None, None, None, None, None, None)
+
+  private val request  = RetrieveHistoricNonFhlUkPiePeriodSummaryRequest(Nino(nino), PeriodId(periodId))
+  private val response = RetrieveHistoricNonFhlUkPiePeriodSummaryResponse(from, to, Some(periodIncome), Some(periodExpenses))
+
+  trait Test extends MockRetrieveHistoricNonFhlUkPropertyPeriodSummaryConnector {
+    implicit val hc: HeaderCarrier              = HeaderCarrier()
+    implicit val logContext: EndpointLogContext = EndpointLogContext("c", "ep")
+
+    val service = new RetrieveHistoricNonFhlUkPropertyPeriodSummaryService(
+      connector = mockRetrieveHistoricNonFhlUkPropertyPeriodSummaryConnector
+    )
+  }
+
+  "retrieve" should {
+    "service call successful" when {
+      "return a valid result" in new Test {
+        MockRetrieveHistoricNonFhlUkPropertyPeriodSummaryConnector
+          .retrieve(request) returns Future.successful(Right(ResponseWrapper(correlationId, response)))
+
+        private val result = await(service.retrieve(request))
+        result shouldBe Right(ResponseWrapper(correlationId, response))
+      }
+    }
+
+    "map errors according to spec" when {
+      def serviceError(downStreamErrorCode: String, error: MtdError): Unit =
+        s"a $downStreamErrorCode error is returned from the service" in new Test {
+
+          MockRetrieveHistoricNonFhlUkPropertyPeriodSummaryConnector
+            .retrieve(request)
+            .returns(Future.successful(Left(ResponseWrapper(correlationId, DownstreamErrors.single(DownstreamErrorCode(downStreamErrorCode))))))
+
+          private val result = await(service.retrieve(request))
+          result shouldBe Left(ErrorWrapper(correlationId, error))
+        }
+
+      val input = Seq(
+        "INVALID_NINO"        -> NinoFormatError,
+        "INVALID_TYPE"        -> InternalError,
+        "INVALID_DATE_FROM"   -> PeriodIdFormatError,
+        "INVALID_DATE_TO"     -> PeriodIdFormatError,
+        "NOT_FOUND_PROPERTY"  -> NotFoundError,
+        "NOT_FOUND_PERIOD"    -> NotFoundError,
+        "SERVER_ERROR"        -> InternalError,
+        "SERVICE_UNAVAILABLE" -> InternalError
+      )
```

Comment:
To avoid duplication of these error codes in both code + tests, can we move these error code map to a companion object or a trait and then either import or extend the trait, as its duplicated in the test code.